api improvements

SimonDanisch opened the pull request:
- add `ignore_equal_values` to Observable constructor, which implements https://github.com/JuliaGizmos/Observables.jl/pull/79 without adding a new type
- add `PriorityObservable` from Makie without adding a new type (from [PriorityObservable.jl](https://github.com/JuliaPlots/Makie.jl/blob/master/src/interaction/PriorityObservable.jl)). One can now simply add `on(f, priority::Int=...)`, to register a callback with priority.
- add `on(f, obs; update=true)`, to call `f(obs[])` immediately.
- improve compile times quite a bit

Based on `sd/compile-time` not for the best reasons, besides it being a bit easier

timholy commented on src/Observables.jl:
```diff
@@ -49,6 +67,31 @@ end
 
+
+"""
+    obs = ChangeObservable(value)
```
Also use a keyword interface?

I don't think I'd capitalize this if it isn't a type.

timholy commented on src/Observables.jl:
```diff
@@ -1,9 +1,9 @@
+export Consume, PriorityObservable, ChangeObservable, ObserverFunction, AbstractObservable
 
 import Base.Iterators.filter
-Base.Experimental.@max_methods 1
```
Is this not useful?

timholy commented on src/Observables.jl:
```diff
@@ -29,16 +29,34 @@ end
+    end
+end
+
+function PriorityObservable(val::T) where T
```
What about making this just `Observable{T}(val; priority=true)`?

timholy commented on src/Observables.jl:
```diff
@@ -64,19 +107,12 @@ end
-    elseif field === :inputs
-        return getfield(obs, field)
-    elseif field === :id
-        return obsid(obs)
```
Was `obsid` already deleted? If so we should remove it from the exports.

timholy commented on src/Observables.jl:
```diff
@@ -64,19 +107,12 @@ end
 end
+Consume() = Consume(true)
+Consume(x::Consume) = Consume(x.x) # for safety in selection_point etc
+Base.:(==)(a::Consume, b::Consume) = a.x == b.x
```
Do you need this method at all? Isn't that the fallback behavior? (It's not `mutable`.)

timholy reviewed (commented):
I'm highly approving of enabling new behaviors without proliferating types. Small comments added.

yha commented:
Some thoughts:
- The boolean `use_priority` field seems unnecessary. If users simply don't pass priority values (resulting in the default priority 0 for everything), and don't use the `Consume` type, you automatically get the "normal" behavior. This redundancy in the interface allows the user to contradict themselves by setting `use_priority=false` and also setting priorities when calling `on` (and the implementation just ignores the priorities and sets them to 0 in this case). I would just remove this field, and the branch in `on`.
- Why is `listeners` a `Vector{Pair{Int,Any}}` and not a `Dict{Int,Any}`?
- Having a boolean `ignore_equal_values` seems unnecessarily restrictive, and it drops the support (currently in `observe_changes`) for specifying the relevant "equality" predicate. That field can more generally be the relevant predicate instead, so it would default to `Returns(false)` (always update), and you can set it to `isequal` to get the current PR behavior, or to other values. I also think this could be a type parameter rather than a field to get more performance. That would mean you can't change it in an existing `Observable`, but is that really needed? (Incidentally, why is `Observable` mutable?)

SimonDanisch commented:
>The boolean use_priority field seems unnecessary

True, good catch! I started with a different implementation where it was needed and forgot to remove now that it isn't needed!

>Why is listeners a Vector{Pair{Int,Any}} and not a Dict{Int,Any}

Because keys occur multiple times! Also, `iterate` + `findfirst` are the main operations we need to use, so vector fits well.

>Having a boolean ignore_equal_values seems unnecessarily restrictive

I wanted to make it as simple as possible for the basic use case, and not make the type more complicated with more type parameters. I'm not 100% sure how much faster 1 type parameter compiles faster than two, but alone the complexity increase to store them concretely in other structs seems not worth it. Especially, since one can easily update with some other comparison method without this flag!

yha commented:
> Because keys occur multiple times!

Right, I meant that a key would map to a vector of listeners. But I realize now that the idea is to maintain a sorted list of keys, so you would actually want a `SortedDict` which is an extra dependency. So a vector of pairs makes sense. But maybe the insertion should be 
```
idx = searchsortedfirst(ls, priority; by=first, rev=true)
insert!(ls, idx, priority => f)
```
since it makes it a bit clearer that it's about maintaining a sorted list? Might just be a matter of taste.



> I'm not 100% sure how much faster 1 type parameter compiles faster than two, but alone the complexity increase to store them concretely in other structs seems not worth it.

I don't follow. Why would you need to store the parameters in other structs? And is the objection to having a predicate instead of a boolean, or just to having as a type parameter?

SimonDanisch commented:
>But maybe the insertion should be

Yes, agreed! :) 

>Why would you need to store the parameters in other structs?

Well, if you want to concretely put any observable in a struct you'll need to do:
```julia
struct OtherStruct{T, COMP}
    obs::Observable{T, COMP}
end
```
I guess one could also do:
```julia
struct OtherStruct{O <: Observable}
    obs::O
end
```
But that doesn't work for all use cases.

timholy commented:
Since the function call is not a type parameter, I don't think the equality check would need to be. But can you give a practical use case where you'd want an equality-comparison operator different from `==`/`isequal`? I'm all in favor of flexiblity if it buys you something, but in this case your choices are (1) make the equality-check field a type-parameter, leading to more specialization (and more latency) for methods that take Observables as arguments, (2) allow it to be `::Any` and thus use runtime dispatch. Hard-coding the comparison operator has neither of these disadvantages, instead it has the disadvantage of being less flexible.

yha commented:
I never used `observe_changes` with a different equality operator, but I can imagine it might be useful with approximate equality, e.g. when tracking mouse movement. (But of course, this can also be done manually on `on`)
My thinking was that using an equality operator is both simpler code and more general, but maybe it's not worth the dynamic dispatch.

yha commented:
Also, I'm not sure `observe_changes` should be dropped. It's still useful as a primitive operation by itself, to do things like
```
map(f, obs1, obs2, observe_changes(obs3))
```
instead of the more cumbersome
```
map(f, obs1, obs2, map(identity, obs3; change_observable=true))
```
I don't know how common this is, though. I've used this pattern, but it's a very small sample.

SimonDanisch commented:
@timholy I went through all `@nospecialize` constructs etc and measured whether they're necessary or not, and removed/added some.

This is now fastest for Makie itself, but let me know if that doesn't work well with some of your Projects!
I think we can merge this now and release a breaking version. I'd like to merge it soon, since we can still tweak a few things in patch releases ;)

I've also brought back most of the features I removed, as long as they didn't have compile time problems.
I'm still removing the weird async features I added at some point, since I don't think anyone has ever used them and I'm not really sure about their implementation.
This PR shaves off around 17% of the ttfp at this point.

timholy commented on docs/src/index.md:
```diff
@@ -43,52 +43,51 @@ obs_func = nothing
+### Priority
 
-#### Delay an update
+One can also give the callback a priority, to enable always calling a specific callback before/after others, independent of the order of registration.
```
Somewhere, one should specify how to interpret the priority: is it always an `Int`? Which occurs first, minimum priority or maximum priority? (Does -1 happen before or after 1?)

timholy commented:
Git history is now really messy, with lots of commit names that aren't very useful (e.g., "does this help??"). Please use squash!
